caom2pipe/caom2pipe/manage_composable.py

```diff
@@ -0,0 +1,725 @@
+
+    def __str__(self):
+        return ' '.join(
+            '{} {}'.format(ii, getattr(self, ii)) for ii in vars(self))
+
+
+class TaskType(Enum):
+    """The possible steps in a Collection pipeline."""
+    STORE = 'store'
+    SCRAPE = 'scrape'
+    INGEST = 'ingest'
+    MODIFY = 'modify'
+    CHECKSUM = 'checksum'
+    VISIT = 'visit'
+
+
+class Config(object):
+    """Configuration information that remains the same for all steps and all
+     work in a pipeline execution."""
+
+    def __init__(self):
+        # the root directory for all executor operations
+        self.working_directory = None
+
+        # the file that contains the list of work to be passed through the
+        # pipeline
+        self.work_file = None
+        # the fully qualified name for the work file
+        self.work_fqn = None
+
+        # credentials for any service calls
+        self.netrc_file = None
+
+        # which collection is addressed by the pipeline
+        self.collection = None
+
+        # changes expectations of the executors for handling files on disk
+        self.use_local_files = False
+
+        # which service instance to use
+        self.resource_id = None
+
+        # the logging level - enforced throughout the pipeline
+        self.logging_level = None
+
+        # write the log to a file?
+        self.log_to_file = False
+
+        # where log files are written to - defaults to working_directory
+        self.log_file_directory = None
+
+        # the ad 'stream' that goes with the collection - use when storing
+        # files
+        self.stream = None
+
+        # the ad 'host' to store files to - used for testing cadc-data put
+        # commands only, should usually be None
+        self.storage_host = None
+
+        # the way to control which steps get executed
+        self.task_types = None
+
+        # the filename where success logs are written
+        # this will be created in log_file_directory
+        self.success_log_file_name = None
+        # the fully qualified name for the file
+        self.success_fqn = None
+
+        # the filename where failure logs are written
+        # this will be created in log_file_directory
+        self.failure_log_file_name = None
+        # the fully qualified name for the file
+        self.failure_fqn = None
+
+        # the filename where retry entries are writter
+        # this will be created in log_file_directory
+        self.retry_file_name = None
+        # the fully qualified name for the file
+        self.retry_fqn = None
+
+        # for the love of pete, don't use this from the config.yml file
+        # it's here for programmatic passing around, and nothing else
+        self.proxy = None
+
+        # feature flags
+        self.features = Features()
+
+    @property
+    def working_directory(self):
+        return self._working_directory
+
+    @working_directory.setter
+    def working_directory(self, value):
+        self._working_directory = value
```

Comment:
The configuration read does not necessarily occur at the same time or on the same machine as when the location needs to be used. This is because of the airflow/k8s/docker deployment environment. So all the existence checks are in the execute methods. If the existence checks are missing from the execute methods, that's a mistake.